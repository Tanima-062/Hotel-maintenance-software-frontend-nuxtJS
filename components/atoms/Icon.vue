<template>
  <div :class="styles">
    <svg
      class="w-7 h-7 fill-current"
      :class="`${color}`"
      :style="`width: ${iconWidth}px; height: ${iconHeight}px`"
      :viewBox="`0 0 24 24`"
    >
      <path :d="icon" />
    </svg>
  </div>
</template>
<script lang="ts">
import { Vue, Component, Prop } from 'nuxt-property-decorator'

@Component
export default class extends Vue {
  @Prop({ default: '' })
  icon!: string

  @Prop({ default: 'text-white' })
  color!: string

  @Prop({ default: '' })
  bgColor!: string

  @Prop({ default: false })
  round!: boolean

  @Prop({ default: 24 })
  iconWidth!: number

  @Prop({ default: 24 })
  iconHeight!: number

  @Prop({ default: '' })
  extraClass!: string

  get styles(): string {
    const classList = []
    if (this.bgColor) {
      classList.push(`${this.bgColor}`)
    }
    if (this.round) {
      classList.push('rounded-full')
    }
    if (this.extraClass) {
      classList.push(this.extraClass)
    }
    return classList.join(' ')
  }
}
</script>
