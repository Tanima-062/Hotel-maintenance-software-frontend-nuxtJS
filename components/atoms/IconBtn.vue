<template>
  <button
    type="button"
    class="flex h-fit-content items-center justify-center"
    :class="styles"
    :disabled="disabled"
    @click="click"
  >
    <svg
      class="w-7 h-7 fill-current"
      :class="`${color} hover:${hoverColor}`"
      :style="`width: ${iconWidth}px; height: ${iconHeight}px`"
      viewBox="0 0 24 24"
    >
      <path :d="icon" />
    </svg>
  </button>
</template>
<script lang="ts">
import { Vue, Component, Prop, Emit } from 'nuxt-property-decorator'

@Component
export default class extends Vue {
  @Prop({ default: '' })
  icon!: string

  @Prop({ default: 'text-white' })
  color!: string

  @Prop({ default: 'text-white' })
  hoverColor!: string

  @Prop({ default: false })
  disabled!: boolean

  @Prop({ default: false })
  round!: boolean

  @Prop({ default: '' })
  bgColor!: string

  @Prop({ default: '' })
  padding!: string

  @Prop({ default: 24 })
  iconWidth!: number

  @Prop({ default: 24 })
  iconHeight!: number

  @Prop({ default: '' })
  extraClass!: string

  private isAnim: boolean = false
  private interval: number | undefined

  get styles(): string {
    const classList = []

    classList.push(this.isAnim ? 'btn--click' : 'btn')
    if (this.disabled) {
      classList.push('cursor-not-allowed')
    }
    if (this.round) {
      classList.push('rounded-full')
    } else if (!this.extraClass.includes('rounded-sm')) {
      classList.push('rounded-md')
    }
    classList.push(this.bgColor ? `${this.bgColor} opacity-75` : 'opacity-85')
    if (this.padding) {
      classList.push(`${this.padding}`)
    }
    if (this.extraClass) {
      classList.push(this.extraClass)
    }
    return classList.join(' ')
  }

  private doAnim(): void {
    if (this.isAnim || this.interval) {
      return
    }
    this.isAnim = true
    this.interval = window.setInterval(() => {
      this.isAnim = false
      clearInterval(this.interval)
      this.interval = undefined
    }, 130)
  }

  @Emit()
  private click(e: Event): Event {
    this.doAnim()
    return e
  }
}
</script>
<style lang="sass" scoped>
.btn
  transition: opacity .1s ease-in
  &:hover
    transition: opacity .2s cubic-bezier(.4,0,.6,1)
    opacity: 1
  &--click
    transition: opacity .05s ease-out
    opacity: .7
.h-fit-content
  height: fit-content
</style>
