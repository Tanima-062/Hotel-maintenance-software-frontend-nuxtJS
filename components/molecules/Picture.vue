<template>
  <transition-group
    tag="div"
    name="picture"
    class="relative bg-black"
    :class="`${width} ${round}`"
  >
    <div
      :key="imageId + '-div'"
      :draggable="draggable"
      :class="`${isHoverPict} ${round}`"
      @dragstart="fDragstart"
      @dragover.prevent="isHover = true"
      @dragleave="isHover = false"
      @dragend="isHover = false"
      @drop.stop="drop"
    >
      <img
        :key="imageId + '-image'"
        :src="imageSrc"
        :class="`${width} ${height} ${round}`"
        draggable="false"
        @error="imageError = true"
        @click.self="clickImage"
      />
    </div>

    <div
      v-if="isImagesCount"
      :key="imageId + '-isImagesCount'"
      :class="countStyle"
    >
      <Icon
        v-if="!showCountTop"
        :icon="icon.image"
        icon-width="12"
        icon-height="12"
        class="mr-1"
      />
      <p
        :class="`text-white ${showCountTop ? 'text-lg' : 'text-sm'} leading-3`"
      >
        {{ imagesCount }}
      </p>
    </div>

    <div
      v-if="isMain"
      :key="imageId + '-isMain'"
      class="absolute top-0 left-0 border-3 border-solid border-alert bg-black bg-opacity-50"
      :class="`${width} ${height} ${round}`"
    >
      <div
        class="bg-alert rounded-br-lg absolute top-0 left-0 text-white py-1 px-2"
        :class="isNotRounded ? '' : 'rounded-br-lg'"
      >
        メイン{{ mainNumber }}
      </div>
    </div>
    <div
      v-if="isChecked"
      :key="imageId + '-isChecked'"
      class="absolute top-0 left-0 border-3 border-solid border-alert"
      :class="`${width} ${height} ${round}`"
    >
      <Icon
        class="absolute top-0 right-0 mr-2 mt-2 p-1"
        :icon="icon.check"
        bg-color="bg-alert"
        icon-width="12"
        icon-height="12"
        :round="true"
      />
    </div>
    <IconBtn
      v-if="isClear && !isViewMode"
      :key="imageId + '-isClear'"
      class="absolute top-0 right-0 mr-2 mt-2"
      :icon="icon.close"
      bg-color="bg-black-dark"
      :round="true"
      @click="clear"
    />
  </transition-group>
</template>
<script lang="ts">
import { Component, Prop, Vue, Emit } from 'nuxt-property-decorator'
import { mdiClose, mdiCheck, mdiImageOutline } from '@mdi/js'

@Component({
  components: {
    IconBtn: () => import('~/components/atoms/IconBtn.vue'),
    Icon: () => import('~/components/atoms/Icon.vue'),
  },
})
export default class extends Vue {
  @Prop({ default: '' })
  imageId!: string

  @Prop({ default: '' })
  href!: string

  @Prop({ default: 'w-60' })
  width!: string

  @Prop({ default: 'h-40' })
  height!: string

  @Prop({ default: false })
  isClear!: boolean

  @Prop({ default: false })
  isMain!: boolean

  @Prop({ default: 1 })
  mainNumber!: number

  @Prop({ default: false })
  isChecked!: boolean

  @Prop({ default: true })
  draggable!: boolean

  @Prop({ default: '' })
  dragDataPrefix!: string

  @Prop({ default: false })
  isNotRounded!: boolean

  @Prop({ default: false })
  isImagesCount!: boolean

  @Prop({ default: 0 })
  imagesCount!: number

  @Prop({ default: false })
  showCountTop!: boolean

  @Prop({ default: false })
  isViewMode!: boolean

  private imageError: boolean = false
  private isHover: boolean = false
  private icon: { [key: string]: string } = {
    close: mdiClose,
    check: mdiCheck,
    image: mdiImageOutline,
  }

  get style(): string {
    const classList = []
    classList.push(this.isHover ? '' : '')
    classList.push(this.width)
    classList.push(this.height)
    return classList.join(' ')
  }

  get countStyle(): string {
    if (this.showCountTop) {
      return 'absolute top-0 left-0 p-2 rounded-br-md rounded-tl-md bg-opacity-50 bg-black-dark'
    } else {
      return 'absolute bottom-0 right-0 mr-3 mb-2 flex items-center justify-cener p-2 rounded-md bg-opacity-50 bg-black-dark'
    }
  }

  get round(): string {
    return this.isNotRounded ? '' : 'rounded-lg'
  }

  get isHoverPict(): string {
    return this.isHover ? 'opacity-50' : ''
  }

  get imageSrc(): string {
    return this.imageError ? '/images/no_image.png' : this.href
  }

  private fDragstart(event: Event & { dataTransfer: DataTransfer }): void {
    event.dataTransfer.setData(
      'text/plain',
      `${this.dragDataPrefix}${this.imageId}`
    )
  }

  @Emit()
  private clear(): string {
    return this.imageId
  }

  @Emit()
  clickImage(): void {}

  @Emit()
  private drop(event: Event & { dataTransfer: DataTransfer }): Object {
    this.isHover = false
    const data = event.dataTransfer.getData('text/plain')

    event.dataTransfer.clearData()
    return {
      from: data,
      to: this.imageId,
    }
  }
}
</script>
<style lang="sass" scoped>
.picture-enter-active
  transition: opacity .2s

.picture-enter, .picture-leave-to
  opacity: 0
</style>
