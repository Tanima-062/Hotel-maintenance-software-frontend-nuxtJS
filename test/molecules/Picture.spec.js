import { mount } from '@vue/test-utils'
import Picture from '@/components/molecules/Picture.vue'
import IconBtn from '@/components/atoms/IconBtn.vue'
import Icon from '@/components/atoms/Icon.vue'

describe('Molecules Picture Check', () => {
  test('Picture Test', () => {
    const wrapper = mount(Picture, {
      stubs: {
        IconBtn,
        Icon,
      },
    })
    // check if it is a vue element
    expect(wrapper.vm).toBeTruthy()

    // check Snapshot
    expect(wrapper.element).toMatchSnapshot()
  })
})
