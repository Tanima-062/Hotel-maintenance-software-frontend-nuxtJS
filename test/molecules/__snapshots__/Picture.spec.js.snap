// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Molecules Picture Check Picture Test 1`] = `
<transition-group-stub
  class="relative bg-black w-60 rounded-lg"
  name="picture"
  tag="div"
>
  <div
    class=" rounded-lg"
    draggable="true"
  >
    <img
      class="w-60 h-40 rounded-lg"
      draggable="false"
      src=""
    />
  </div>
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
</transition-group-stub>
`;
